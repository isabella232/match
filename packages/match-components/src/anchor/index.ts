export * from "./anchor";
export * from "./constants";
export * from "./types";
