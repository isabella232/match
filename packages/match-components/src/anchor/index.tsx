import * as React from "react";
import * as PropTypes from "prop-types";
import { StyledAnchor } from "./styles";
import { AnchorVariant, AnchorProps } from "./types";

export interface ExternalProps {
  rel: string | undefined;
  target: string | undefined;
}

const EXTERNAL_URL_REGEX = /^(https?:)\S*$/;

const secureExternalLink = (href: string | undefined): ExternalProps => {
  if (href) {
    if (EXTERNAL_URL_REGEX.test(href)) {
      return { rel: "noreferrer noopener", target: "_blank" };
    }
  }
  return { rel: undefined, target: undefined };
};

const Anchor: React.FC<AnchorProps> = ({ ...props }) => {
  if (!props.href) console.warn("href must be provided for an anchor");
  const externalURL = secureExternalLink(props.href);
  return (
    <StyledAnchor
      {...props}
      rel={props.rel ? props.rel : externalURL.rel}
      target={props.target ? props.target : externalURL.target}
    />
  );
};

Anchor.propTypes = {
  variant: PropTypes.oneOf(Object.values(AnchorVariant)),
  href: PropTypes.string,
  target: PropTypes.string,
  rel: PropTypes.string,
  noUnderline: PropTypes.bool,
};

Anchor.defaultProps = {
  variant: AnchorVariant.TEXT,
  noUnderline: false,
};

export { Anchor, AnchorVariant, AnchorProps };
