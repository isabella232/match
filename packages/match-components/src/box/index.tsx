import * as React from "react";
import { spaceTokenProp } from "@twilio-labs/match-props";
import { BoxProps } from "./types";
import { StyledBox } from "./styles";

const Box: React.FC<BoxProps> = ({ ...props }) => {
  return <StyledBox {...props} />;
};

Box.propTypes = {
  padding: spaceTokenProp,
  paddingLeft: spaceTokenProp,
  paddingRight: spaceTokenProp,
  paddingBottom: spaceTokenProp,
  paddingTop: spaceTokenProp,
  margin: spaceTokenProp,
  marginLeft: spaceTokenProp,
  marginRight: spaceTokenProp,
  marginBottom: spaceTokenProp,
  marginTop: spaceTokenProp,
};

export { Box, BoxProps };
