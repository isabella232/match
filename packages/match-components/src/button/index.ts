export * from "./constants";
export * from "./button";
export * from "./types";
