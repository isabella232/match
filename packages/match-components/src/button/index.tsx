import * as React from "react";
import * as PropTypes from "prop-types";
import { uid } from "react-uid";
import { StyledButton, StyledPrompt } from "./styles";
import { ButtonVariant, ButtonType, ButtonSize } from "./types";
import type { ButtonProps } from "./types";

const Button: React.FC<ButtonProps> = ({
  type,
  prompt,
  children,
  ...props
}) => {
  if (props.download && !props.href)
    console.warn("[Button]: Href must be provided for a download link.");

  return (
    <StyledButton
      forwardedAs={props.href ? "a" : undefined}
      type={!props.href ? type : undefined}
      {...props}
    >
      {Array.isArray(children)
        ? children.map((child) => {
            if (React.isValidElement(child))
              return React.cloneElement(child, { key: uid(child) });
            if (typeof child === "string") return child.trim();
          })
        : children}
      {prompt && <StyledPrompt />}
    </StyledButton>
  );
};

Button.propTypes = {
  variant: PropTypes.oneOf(Object.values(ButtonVariant)),
  type: PropTypes.oneOf(Object.values(ButtonType)),
  size: PropTypes.oneOf(Object.values(ButtonSize)),
  href: PropTypes.string,
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  download: PropTypes.bool,
  prompt: PropTypes.bool,
};

Button.defaultProps = {
  variant: ButtonVariant.PRIMARY,
  type: ButtonType.BUTTON,
  size: ButtonSize.NORMAL,
  disabled: false,
  fullWidth: false,
  download: false,
  prompt: false,
};

Button.displayName = "Button";

export { Button, ButtonVariant, ButtonType, ButtonSize };
export type { ButtonProps };
