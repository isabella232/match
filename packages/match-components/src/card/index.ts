export * from "./card";
export * from "./constants";
export * from "./types";
