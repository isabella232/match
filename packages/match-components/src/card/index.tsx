import * as React from "react";
import { spaceTokenProp } from "@twilio-labs/match-props";
import { CardProps, CardVariant } from "./types";
import { StyledCard } from "./styles";
import * as PropTypes from "prop-types";

const Card: React.FunctionComponent<CardProps> = ({ ...props }) => {
  return <StyledCard {...props} />;
};

Card.propTypes = {
  variant: PropTypes.oneOf(Object.values(CardVariant)),
  padding: spaceTokenProp,
  paddingLeft: spaceTokenProp,
  paddingRight: spaceTokenProp,
  paddingBottom: spaceTokenProp,
  paddingTop: spaceTokenProp,
  margin: spaceTokenProp,
  marginLeft: spaceTokenProp,
  marginRight: spaceTokenProp,
  marginBottom: spaceTokenProp,
  marginTop: spaceTokenProp,
};

Card.defaultProps = {
  variant: CardVariant.PRIMARY,
};

export { Card, CardProps };
