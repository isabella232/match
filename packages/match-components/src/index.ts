export * from "./anchor";
export * from "./button";
export * from "./card";
export * from "./code";
export * from "./form";
export * from "./grid";
export * from "./heading";
export * from "./input";
export * from "./paragraph";
export * from "./richtext";
export * from "./separator";
export * from "./snippet";
export * from "./textarea";
export * from "./truncate";
export * from "./visually-hidden";
