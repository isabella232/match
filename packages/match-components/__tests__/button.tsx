import * as React from "react";
import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import { withTheme } from "@twilio-labs/match-themes";
import { Button } from "../src";

const ButtonWithTheme = withTheme()(Button);

describe("Button", () => {
  test("renders as an anchor", () => {
    const { getByText } = render(
      <ButtonWithTheme href="https://twilio.com">Click Me</ButtonWithTheme>
    );
    expect(getByText(/click me/i).tagName.toLowerCase()).toBe("a");
  });

  test("renders as a button", () => {
    const { getByText } = render(<ButtonWithTheme>Click Me</ButtonWithTheme>);
    expect(getByText(/click me/i).tagName.toLowerCase()).toBe("button");
  });

  test("snapshot", (): void => {
    const { asFragment } = render(<ButtonWithTheme>Click Me</ButtonWithTheme>);
    expect(asFragment()).toMatchSnapshot();
  });

  test("accessibility violations", async () => {
    const { container } = render(<ButtonWithTheme>Click Me</ButtonWithTheme>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
