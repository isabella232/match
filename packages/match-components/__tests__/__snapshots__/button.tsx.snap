// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot 1`] = `
.c0 {
  color: hsla(217.33333333333334,42.05607476635513%,20.980392156862745%,1);
  font-weight: 400;
  font-size: 1rem;
  font-family: Whitney SSm A,Whitney SSm B,Helvetica Neue,Helvetica,Arial,sans-serif;
}

.c1 {
  display: inline-block;
  font-weight: 500;
  font-family: Whitney SSm A,Whitney SSm B,Helvetica Neue,Helvetica,Arial,sans-serif;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-style: solid;
  border-width: 2px;
  border-radius: 4px;
  cursor: pointer;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  -webkit-transition-property: color,background,border;
  transition-property: color,background,border;
  -webkit-transition-timing-function: ease-in;
  transition-timing-function: ease-in;
  padding: calc(16px - 2px) calc(24px - 2px);
  font-size: 1rem;
  color: hsla(0,0%,100%,1);
  background: hsla(213.78378378378378,98.23008849557523%,44.31372549019608%,1);
  border-color: hsla(213.78378378378378,98.23008849557523%,44.31372549019608%,1);
}

.c1:focus {
  outline-width: 2px;
  outline-style: solid;
  outline-color: hsla(213.78378378378378,98.23008849557523%,44.31372549019608%,1);
  outline-offset: 2px;
  -moz-outline-radius: 4px;
}

.c1:disabled {
  color: hsla(219.5121951219512,26.451612903225808%,30.3921568627451%,1);
  background: hsla(240,9.999999999999986%,96.07843137254902%,1);
  border-color: hsla(240,9.999999999999986%,96.07843137254902%,1);
  cursor: not-allowed;
}

.c1:hover {
  background: hsla(221.01694915254237,95.67567567567566%,36.274509803921575%,1);
  border-color: hsla(221.01694915254237,95.67567567567566%,36.274509803921575%,1);
}

.c1:focus,
.c1:active {
  background: hsla(231.24087591240877,100%,26.862745098039216%,1);
  border-color: hsla(231.24087591240877,100%,26.862745098039216%,1);
}

<div
  class="c0"
>
  <button
    class="c1"
    type="button"
  >
    Click Me
  </button>
</div>
`;
